src/imp/schannel.rs

```diff
@@ -140,15 +154,122 @@ impl Identity {
+
+                    let algo: HashAlgorithm; 
+                    match decoded_hex.len() {
+                        20 => algo = HashAlgorithm::sha1(), 
+                        32 => algo = HashAlgorithm::sha256(), 
+                        _ => return Err(io::Error::new(io::ErrorKind::InvalidInput,"Invalid hex thumbprint").into())
+                    }
+
+                    let hash = cert.fingerprint(algo)?; 
+                    if hash == decoded_hex { 
+                        if cert 
+                            .private_key() 
+                            .silent(true) 
+                            .compare_key(true) 
+                            .acquire() 
+                            .is_err() 
+                        { 
+                            return Err(io::Error::new(io::ErrorKind::InvalidInput,"Missing or invalid private key property").into())
+                        } 
+                        identity = Some(cert);
+                        break; 
+                    } 
+                } 
+
+                let identity = match identity {
+                    Some(identity) => identity,
+                    None => {
+                        return Err(io::Error::new(io::ErrorKind::InvalidInput,"No identity found in provided store").into());
+                    }
+                };
+                return Ok(Identity { cert: identity })
+            }
+
+            OsProviderParameters::ContextFromFile {file_path} => {
+                let store = CertStore::open_file(&file_path.as_path())?;
+                // set identity to the cert that matches the key inside the pfx file
+                let mut identity = None;
+                for cert in store.certs() {
+                    if cert
+                        .private_key()
+                        .silent(true)
+                        .compare_key(true)
+                        .acquire()
+                        .is_ok()
+                    {
+                        identity = Some(cert);
+                        break;
+                    }
+                }
+                let identity = match identity {
+                    Some(identity) => identity,
+                    None => {
+                        return Err(io::Error::new(io::ErrorKind::InvalidInput,"No identity found in provided store").into());
+                    }
+                };
+                return Ok(Identity { cert: identity })
+            }
+        }
+    }
 }
 
+
 // The name of the container must be unique to have multiple active keys.
 fn gen_container_name() -> String {
     use std::sync::atomic::{AtomicUsize, Ordering};
     static COUNTER: AtomicUsize = AtomicUsize::new(0);
     format!("native-tls-{}", COUNTER.fetch_add(1, Ordering::Relaxed))
 }
 
+fn parse_engine_string(engine_string: &OsStr) -> io::Result<OsProviderParameters> { 
+
+    let converted_str = engine_string.to_string_lossy(); // convert back to to_string, hanlde error
```

Comment:
>to_string_lossy [](http://example.com/codeflow?start=38&length=15)

change to to_str. For None, return Err, with "Contains invalid UTF-8 characters.